cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AppLifecycleHttpHandler.java

```
@@ -378,6 +382,70 @@ public void updateApp(FullHttpRequest request, HttpResponder responder,
     }
   }
 
+  /**
+   * upgrades an existing application.
+   */
+  @POST
+  @Path("/apps/{app-id}/upgrade")
+  @AuditPolicy(AuditDetail.REQUEST_BODY)
+  public void upgradeApplication(HttpRequest request, HttpResponder responder,
+                                 @PathParam("namespace-id") final String namespaceId,
+                                 @PathParam("app-id") final String appName)
+      throws NotFoundException, BadRequestException, UnauthorizedException, IOException {
+
+    ApplicationId appId = validateApplicationId(namespaceId, appName);
+
+    try {
+      ApplicationUpdateDetails detail =
+          applicationLifecycleService.upgradeApplication(appId, createProgramTerminator());
+      responder.sendJson(HttpResponseStatus.OK, GSON.toJson(detail));
+    } catch (Exception e) {
+      LOG.error("Upgrade application failure", e);
+      ApplicationUpdateDetails detail = new ApplicationUpdateDetails(
+          new ServiceException("Upgrade failed due to internal error.", null,
```

Does upgradeApplication() throw NotFoundException, BadRequestException, etc? If so, this isn't guaranteed to be an internal error. 

It seems like you can just let the exception propagate and not catch it here at all.